backend/controller/evaljob_controller.go

```diff
@@ -0,0 +1,568 @@
+			handler.EnqueueRequestForOwner(mgr.GetScheme(), mgr.GetRESTMapper(), &lmevalservicev1beta1.EvalJob{}),
+			builder.WithPredicates(predicate.Funcs{
+				// drop all events except deletion
+				CreateFunc: func(event.CreateEvent) bool {
+					return false
+				},
+				UpdateFunc: func(event.UpdateEvent) bool {
+					return false
+				},
+				GenericFunc: func(event.GenericEvent) bool {
+					return false
+				},
+			}),
+		).
+		Complete(r)
+}
+
+func (r *EvalJobReconciler) constructOptionsFromConfigMap(configmap *corev1.ConfigMap) error {
+	r.options.DriverImage = DefaultDriverImage
+	r.options.PodImage = DefaultPodImage
+	r.options.DriverServiceAccount = DefaultDriverServiceAccount
+	if v, found := configmap.Data[DriverImageKey]; found {
+		r.options.DriverImage = v
+	}
+	if v, found := configmap.Data[PodImageKey]; found {
+		r.options.PodImage = v
+	}
+	if v, found := configmap.Data[DriverServiceAccountKey]; found {
+		r.options.DriverServiceAccount = v
+	}
+	return nil
+}
+
+func (r *EvalJobReconciler) handleDeletion(ctx context.Context, job *lmevalservicev1beta1.EvalJob, log logr.Logger) (reconcile.Result, error) {
+	if controllerutil.ContainsFinalizer(job, lmevalservicev1beta1.FinalizerName) {
+		// delete the correspondling pod if needed
+		// remove our finalizer from the list and update it.
+		if job.Status.State != lmevalservicev1beta1.CompleteJobState ||
+			job.Status.Reason != lmevalservicev1beta1.CancelledReason {
+
+			if err := r.deleteJobPod(ctx, job); err != nil {
+				log.Error(err, "failed to delete pod of the job")
+			}
+		}
+
+		controllerutil.RemoveFinalizer(job, lmevalservicev1beta1.FinalizerName)
+		if err := r.Update(ctx, job); err != nil {
+			return ctrl.Result{}, err
+		}
+		r.Recorder.Event(job, "Normal", "DetachFinalizer",
+			fmt.Sprintf("removed finalizer from EvalJob %s in namespace %s",
+				job.Name,
+				job.Namespace))
+		log.Info("Successfully remove the finalizer", "name", job.Name)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *EvalJobReconciler) handleNewCR(ctx context.Context, log logr.Logger, job *lmevalservicev1beta1.EvalJob) (reconcile.Result, error) {
+	// If it doesn't contain our finalizer, add it
+	if !controllerutil.ContainsFinalizer(job, lmevalservicev1beta1.FinalizerName) {
+		controllerutil.AddFinalizer(job, lmevalservicev1beta1.FinalizerName)
+		if err := r.Update(ctx, job); err != nil {
+			log.Error(err, "unable to update finalizer")
+			return ctrl.Result{}, err
+		}
+		r.Recorder.Event(job, "Normal", "AttachFinalizer",
+			fmt.Sprintf("added the finalizer to the EvalJob %s in namespace %s",
+				job.Name,
+				job.Namespace))
+		// Since finalizers were updated. Need to fetch the new EvalJob
+		// End the current reconsile and get revisioned job in next reconsile
+		return ctrl.Result{}, nil
+	}
+
+	// construct a new pod and create a pod for the job
+	currentTime := v1.Now()
+	pod := r.createPod(job)
+	if err := r.Create(ctx, pod, &client.CreateOptions{}); err != nil {
```

Comment:
I'm not super familiar with how `kubebuilder` controllers get wired into the operator process. Is there any chance of a race here where the check to see if it's a new CR says "yes," but by the time this `r.Create` happens it's already been reconciled by another reconciliation process or a sibling operator? I suppose there's always the race that a user might manually make the pod themselves, but we probably don't need to account for that.

Reply:
ha, you are talking about the scaling of the controller now. for scaling, each controller should work on its own bucket and different CRs should go to different buckets. At least that's how knative works. I will check on how multiple replica of the `kubebuilder` controller works. For the racing condition, theoretically, one reconciliation would throw an error and the other reconciliation would pass. So we can have a logic to validate the pod status to correct the state of the CR. I noted this down and will add some logic to handle that.